### PR TITLE
Basic state manager

### PR DESCRIPTION
**Pull request by maximeLong**

### CODE REVIEW

File: Assets/HotmaxEditorVR/stateManager.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        xRotationMode,
+        yRotationMode,
+        zRotationMode
+    }
+    public delegate void rotationModeHandler(rotationModes value);
+    public static event rotationModeHandler rotationModeEvent;
+
+
+
+
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/stateManager.cs]: clean up spacing, make sure vars are with their delegate and event without break

File: Assets/HotmaxEditorVR/stateManager.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    public static event selectedObjectHandler selectedObjectEvent;
+
+
+
+    public static bool selectedObjectIsActive = false;
+
+    public delegate void selectedObjectIsActiveHandler(bool value);
+    public static event selectedObjectIsActiveHandler selectedObjectIsActiveEvent;
+
+    public static bool rotationGizmoIsSelected = false;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/stateManager.cs]: put rotation stuff together

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 3]
```diff
@@ -0,0 +1,223 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class universalTransform : MonoBehaviour
+{
+    public objectSelect objSelect;
+
+    float initialPadYPosition;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: be explicit with declaring private variables -- also would help to group variables together that are related and add a comment above the group that briefly states functionality

Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/universalTransform.cs]: I see what you mean, you want to have it "private" before

**Comment by maximeLong** [comment]
Put private before variable name if it's private.

File: Assets/HotmaxEditorVR/stateManager.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public enum editorModes
+    {
+        universalTransformMode,
+        cloneDeleteMode,
+        openMenuMode
+    }
+    public delegate void editorModeHandler(editorModes value);
+    public static event editorModeHandler editorModeEvent;
+
+    public static bool rotationGizmoIsSelected = false;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/stateManager.cs:17]: can you clean up these spaces, make sure everything in the top is grouped together

File: Assets/HotmaxEditorVR/cloneControl.cs [hunk 5]
```diff
@@ -4,38 +4,59 @@
...
+
+    void updateSelectedObjectEvent(GameObject value)
+    {
+        _selectedObject = value;
+    }
+
+    //clone the currently selected object and delete the editing behaviours from it
+    void cloneSelectedObject()
+    {
+        if (_editorMode == stateManager.editorModes.cloneDeleteMode)
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/cloneControl.cs]: You should do any checks like this outside of the function - so  on line 36 is a better place. Basically a function should run only when it actually needs to.

File: Assets/HotmaxEditorVR/inputManager.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        //Not sure why steamVR's select hand option doesn't do this, let me know if you figure it out
+        hand1 = GameObject.Find("Hand1").GetComponent<Valve.VR.InteractionSystem.Hand>();
+        hand2 = GameObject.Find("Hand2").GetComponent<Valve.VR.InteractionSystem.Hand>();
+        trackedController1 = hand1.gameObject.GetComponent<SteamVR_TrackedController>();
+        trackedController2 = hand2.gameObject.GetComponent<SteamVR_TrackedController>();
+        trackedObject2 = hand2.GetComponent<SteamVR_TrackedObject>();
+        device2 = SteamVR_Controller.Input((int)trackedObject2.index);
+
+
+        Invoke("setControllerIndecies", 1.2f);
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/inputManager.cs]: whats the deal with invoke?

File: Assets/HotmaxEditorVR/inputManager.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        stateManager.editorModeEvent -= updateEditorMode;
+
+    }
+
+    void updateSelectedObjectIsActive(bool value)
+    {
+        _selectedObjectIsActive = value;
+    }
+
+    void updateEditorMode(stateManager.editorModes value)
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/inputManager.cs]: remove this and the listeners if we dont need them right now

File: Assets/HotmaxEditorVR/inputManager.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+public class inputManager : MonoBehaviour
+{
+    // -- controller refs
+    public static Valve.VR.InteractionSystem.Hand hand1;
+    public static Valve.VR.InteractionSystem.Hand hand2;
+    public static SteamVR_TrackedController trackedController1;
+    public static SteamVR_TrackedController trackedController2;
+    private SteamVR_TrackedObject trackedObject2;
+    public static SteamVR_Controller.Device device2;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/inputManager.cs]: can we call device2 something else more descriptive? This is going to be referenced a lot in this software. And do a ctrl+alt+F and change everywhere

File: Assets/HotmaxEditorVR/objectSelect.cs [hunk 9]
```diff
@@ -5,174 +5,158 @@
 
 public class objectSelect : MonoBehaviour
 {
-    public Valve.VR.InteractionSystem.Hand hand1;
-    public Valve.VR.InteractionSystem.Hand hand2;
+    //--laser stuff
+    private Vector3 laserEndPosition;
+    public LineRenderer laserLineRenderer;
+    private float laserWidth = 0.01f;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/objectSelect.cs]: make 11 and 12 private, but add [SerializeField] for inspector

File: Assets/HotmaxEditorVR/objectSelect.cs [hunk 10]
```diff
@@ -5,174 +5,158 @@
 
 public class objectSelect : MonoBehaviour
 {
-    public Valve.VR.InteractionSystem.Hand hand1;
-    public Valve.VR.InteractionSystem.Hand hand2;
+    //--laser stuff
+    private Vector3 laserEndPosition;
+    public LineRenderer laserLineRenderer;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/objectSelect.cs]: make 10 private?

File: Assets/HotmaxEditorVR/objectSelect.cs [hunk 11]
```diff
@@ -5,174 +5,158 @@
...
-               
-                if (hit.collider.name.Contains("(1)"))
-                {
-                    GameObject spawn = Instantiate(Resources.Load("(1)", typeof(GameObject))) as GameObject;
-                    spawn.transform.position = hit.collider.transform.position;
-                    spawn.transform.rotation = hit.collider.transform.rotation;
-                }
-                if (hit.collider.name.Contains("(2)"))
+                // selected object selection
+                if (hit.collider.gameObject == _selectedObject)
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/objectSelect.cs]: This whole raycast check system needs to be smarter. 
The best way right now is create an Interface for raycastable stuff, extend anything that is raycastable with that interface, and then do a single check for typeOf raycastable -- and those classes should at that point handle a raycast hit however makes sense to them.
We'll go over this tomorrow, we'll need to study a little bit. I can see this getting out of hand if we don't tackle it now.

Comment (maximeLong) [reply Assets/HotmaxEditorVR/objectSelect.cs]: https://unity3d.com/learn/tutorials/topics/scripting/interfaces

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+using UnityEngine;
+
+public class universalTransform : MonoBehaviour
+{
+    private float _initialPadYPosition;
+    private float currentPadYPos;
+
+    private float distToController;
+    private float scrollSpeed = .2f;
+    public float scrollDistance = 0;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: private var

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 13]
```diff
@@ -0,0 +1,225 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class universalTransform : MonoBehaviour
+{
+    private float _initialPadYPosition;
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: remove underscore

File: Assets/HotmaxEditorVR/universalTransform.cs (hunk 12 above)
Comment (maximeLong) [reply Assets/HotmaxEditorVR/universalTransform.cs]: and [SerializeField] items on 11

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+            }
+        }
+    }
+
+    void updateEditorMode(stateManager.editorModes value)
+    {
+        _editorMode = value;
+    }
+
+    void setOneRotationGizmoVisible(int gizmo)
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: condense setOneRotationGizmoVisible and setAllRotationsGizmosVisible to one function, that takes an enum argument - with NONE, X, Y, Z -- and switch statement with functionality

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+                    scrollDistance += scrollSpeed;
+                    _initialPadYPosition = currentPadYPos;
+                }
+                if (currentPadYPos < _initialPadYPosition)
+                {
+                    scrollDistance -= scrollSpeed;
+                    _initialPadYPosition = currentPadYPos;
+                }
+            }
+            else
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: move ELSE block after both IF blocks

File: Assets/HotmaxEditorVR/universalTransform.cs [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+            }
+
+            //Rotation Mode
+            if (_rotationGizmoIsSelected)
+            {
+                currentXControllerPosition = inputManager.hand2.transform.localPosition.x;
+                currentYControllerPosition = inputManager.hand2.transform.localPosition.y;
+                float moveBuffer = 0.01f;
+
+                if (_rotationMode == stateManager.rotationModes.xRotationMode)
```
Comment (maximeLong) [inline Assets/HotmaxEditorVR/universalTransform.cs]: You are planning on redoing this work with another system, right? 
There is a lot of redundancy here that could be shortened if anything were to stay.

File: Assets/HotmaxEditorVR/cloneControl.cs (hunk 5 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/cloneControl.cs]: got it.

File: Assets/HotmaxEditorVR/inputManager.cs (hunk 6 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/inputManager.cs]: It is waiting for a steamVR method which fires 1s after the app starts.  The purpose of the steamVR method is to find the hardware indices of the controllers.

File: Assets/HotmaxEditorVR/inputManager.cs (hunk 8 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/inputManager.cs]: yeah, I can call it selectorHand or rightHand

Comment (maximeLong) [reply Assets/HotmaxEditorVR/inputManager.cs]: selectorDevice would be better actually

File: Assets/HotmaxEditorVR/inputManager.cs (hunk 7 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/inputManager.cs]: got it!

File: Assets/HotmaxEditorVR/objectSelect.cs (hunk 9 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/objectSelect.cs]: got it!

File: Assets/HotmaxEditorVR/objectSelect.cs (hunk 10 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/objectSelect.cs]: it has to be public so other classes can grab it. we are using the hand2 in "universalTransform" to fetch the gameobjects worldspace transform.  We won't be doing this in the future. maybe we can change it to private later. ill put a todo there

File: Assets/HotmaxEditorVR/universalTransform.cs (hunk 12 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/universalTransform.cs]: got it!

File: Assets/HotmaxEditorVR/universalTransform.cs (hunk 14 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/universalTransform.cs]: Got it!

File: Assets/HotmaxEditorVR/universalTransform.cs (hunk 15 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/universalTransform.cs]: I will wrap these in their own functions so the if else stuff is separate and more readable.

File: Assets/HotmaxEditorVR/universalTransform.cs (hunk 16 above)
Comment (Andy-Roger) [reply Assets/HotmaxEditorVR/universalTransform.cs]: will refine the updated code after push

**Pull request merged**